libsplinter/src/network/peer_manager/mod.rs

```diff
@@ -0,0 +1,1040 @@
+}
+
+impl PeerRef {
+    pub(super) fn new(peer_id: String, peer_remover: PeerRemover) -> Self {
+        PeerRef {
+            peer_id,
+            peer_remover,
+        }
+    }
+
+    pub fn peer_id(&mut self) -> &str {
+        &self.peer_id
+    }
+}
+
+impl Drop for PeerRef {
+    fn drop(&mut self) {
+        match self.peer_remover.remove_peer_ref(&self.peer_id) {
+            Ok(_) => (),
+            Err(err) => error!(
+                "Unable to remove reference to {} on drop: {}",
+                self.peer_id, err
+            ),
+        }
+    }
+}
+
+/// The PeerManager is in charge of keeping track of peers and their ref count, as well as
+/// requesting connections from the ConnectionManager. If a peer has disconnected, the PeerManager
+/// will also try the peer's other endpoints until one is successful.
+pub struct PeerManager {
+    connection_manager_connector: Connector,
+    join_handle: Option<thread::JoinHandle<()>>,
+    sender: Option<Sender<PeerManagerMessage>>,
+    shutdown_handle: Option<ShutdownHandle>,
+    max_retry_attempts: Option<u64>,
+}
+
+impl PeerManager {
+    pub fn new(connector: Connector, max_retry_attempts: Option<u64>) -> Self {
+        PeerManager {
+            connection_manager_connector: connector,
+            join_handle: None,
+            sender: None,
+            shutdown_handle: None,
+            max_retry_attempts,
+        }
+    }
+
+    /// Start the PeerManager
+    ///
+    /// Returns a PeerManagerConnector that can be used to send requests to the PeerManager.
+    pub fn start(&mut self) -> Result<PeerManagerConnector, PeerManagerError> {
+        let (sender, recv) = channel();
+        if self.sender.is_some() {
+            return Err(PeerManagerError::StartUpError(
+                "PeerManager has already been started".to_string(),
+            ));
+        }
+        let connector = self.connection_manager_connector.clone();
+        let peer_remover = PeerRemover {
+            sender: sender.clone(),
+        };
+
+        let mut subscriber = connector.subscribe().expect("Cannot get subscriber");
+        let notification_sender = sender.clone();
+        // start thread to listen for notifications from the connection manager
+        thread::Builder::new()
+            .name("Peer Manager Notification".into())
+            .spawn(move || loop {
+                // convert connection mangager notification to peer manager notifications
+                // this enables it to be handled in the same thread as other peer manager messages
+                match subscriber.next() {
+                    Some(notification) => {
+                        match notification_sender
+                            .send(PeerManagerMessage::InternalNotification(notification))
+                        {
+                            Ok(_) => (),
+                            Err(_) => {
+                                error!("Peer Manager has shutdown");
```

Comment:
Are these actually error states?

Reply:
I guess not, what log level should this be?

Reply:
Changed them to info